doc: Add core logic of finality provider and status transition (#71)
babylonlabs-io/finality-provider: docs/fp-core.md

# Finality Provider Core

This doc is to provide basic invariants and heuristics one should follow to
build the finality provider. This part should be agnostic to any consumer
chains.

## Status Transitions

> The lifecycle of a finality provider instance involves 5 status: `Registered`,
> `Inactive`, `Active`, `Jailed`, `Slashed`. The status transition is depicted

Reviewer: ```suggestion
A finality provider has the following 5 possible states: `Registered`,
`Inactive`, `Active`, `Jailed`, `Slashed`. The state transition is depicted
```

in the following diagram.

![Finality Provider Status Transition](./fp-status-transition.png).

## Invariants

We define invariants that the finality provider should satisfy:

> - Finality signature should be submitted for a block if:

Reviewer: ```suggestion
- The finality provider should submit a finality signature for a block if:
```

  - the block is not finalized,
  - the finality provider has positive voting power for this block,
  - the finality provider has not cast a finality signature for this block, and
  - the finality provider has cast finality votes for all non-finalized blocks
    prior to this block for which it has voting power.

> - Public randomness should be committed for a block if:

Reviewer: ```suggestion
- The finality provider should commit public randomness for a block if:
```

>   - the block is not finalized, and

Reviewer: Is this constraint necessary and did we implement it?

Author: This indicates that if the block is already finalized, we can skip it but it does not mean that we cannot commit pub rand for a finalized block. We have implemented in https://github.com/babylonlabs-io/finality-provider/blob/8c63dae68284cb5426cb9cc07967e367c7d627cd/finality-provider/service/fp_instance.go#L502

  - the finality provider has not committed any randomness for this block

- A finality provider instance should not be running if it’s status is `slashed`

>   or `jailed` on Babylon.

Reviewer: Is it that the FP should not be running or this FP is still running but somehow remains silent?

Author: we have a monitor to check status of the running fp. if the status is detected to be `jailed` and `slashed`, we should stop the instance as it will be a waste of resources to run it. The operator should identify and fix the issue immediately. See our current impl https://github.com/babylonlabs-io/finality-provider/blob/8c63dae68284cb5426cb9cc07967e367c7d627cd/finality-provider/service/fp_manager.go#L130

> #### Internal variables

Reviewer: ```suggestion
### Internal variables
```

The finality provider maintains the following internal variables.
* `last_voted_height`: the last block height on which the fp has cast a vote,
* `last_processed_height`: the last block height on which the fp either has
  cast a vote or it does not have voting power, and

> * `next_height`: the next block to poll.

Reviewer: Do you think we should distinguish between internal variables of the FP instance and the poller?

Author: As there are not many internal variables to maintain, I think we can just list all of them here.

Internal variables need to satisfy the following rules:
* `last_processed_height >= last_voted_height` should always hold, and

> * `next_height > max{latest_finalized_height, last_processed_height}` should
>   always hold.

Reviewer: Is 
```suggestion
* `next_height = max{latest_finalized_height, last_processed_height} + 1` should
  always hold.
```

always true?

Author: Good point

## Heuristics

### Sending finality votes

> * The poller polls blocks from start_height one by one monotonically.

Reviewer: ```suggestion
* The poller polls blocks from `start_height` one by one monotonically.
```

### Committing public randomness

* Finality providers should start to send public randomness commit right
  after it is registered on Babylon. This is because a public randomness commit
  only takes into effect after the committed epoch is finalized. So the finality
  provider should start sending the commit as soon as possible.
* If the remaining randomness is not sufficient (defined by config) by the
  current babylon height, the finality provider should make a new commit.
* The next commit should be made with `start_height` right after the last
  committed height. This is to ensure no gap between each commit.

The above approach of committing public randomness may cause waste of some
public randomness in cases where the finality provider does not have voting
power for some heights, but given that the randomness are constructed in a
merkle-tree and the commit only contains the root hash of the tree. Therefore,
the waste is negligible.

### Fast sync

Fast sync is needed for bootstrap and in case the finality provider falls
behind due to network latency.

* The finality provider enters fast sync if:
  * the finality provider is bootstrapping, or

>   * the `current_block_height - the last processed height > gap` (configured value).

Reviewer: After https://github.com/babylonlabs-io/finality-provider/pull/57, FP in the normal-case also batch submits finalkity signatures. Does it mean that the normal-case execution is same as fast sync with smaller gaps? If yes, does it make sense to harmoise the normal-case execution and fastsync, or even remove the fast sync completely?

Author: Possibly we can remove the case where the fp enters fast-sync from normal-execution but we still need fast sync for bootstrap. I'll keep it for now

Reviewer: I see, sg
